### PR TITLE
Add a post about how fast CFG generation can be

### PR DESCRIPTION
**Pull request by brandonwillard**
This PR adds an `org-mode` file containing exposition on the basic latency of CFG-guided generation using our approach.

- [x] Determine how much of the interface we want to show (e.g. include/exclude all the index stuff, don't show `ParserLogitsProcessor`, etc.)

### CODE REVIEW

File: org/how-fast-cfg.org [hunk 1]
```diff
@@ -0,0 +1,778 @@
...
+
+np.std(cfg_times)
+# 0.00022955151532709527
+
+%timeit cfg_generator(prompt, max_tokens=100, rng=rng)
+# 2.25 s ± 621 ms per loop (mean ± std. dev. of 7 runs, 1 loop each)
+#+END_SRC
+
+As we can see, our CFG-structured generation adds on average sub-millisecond
+latencies, and it barely breaks a millisecond at maximum.
```
Comment (quernd) [inline org/how-fast-cfg.org:376]: Can't we make an even stronger claim? The overhead is not zero either for unstructured generation, or am I reading the numbers wrong?

File: org/how-fast-cfg.org [hunk 2]
```diff
@@ -0,0 +1,778 @@
...
+res = llm(
+    prompt,
+    max_tokens=100,
+    seed=2305,
+    logits_processor=LogitsProcessorList([logits_processor]),
+)
+# Llama.generate: prefix-match hit
+#
+# llama_print_timings:        load time =    1172.23 ms
+# llama_print_timings:      sample time =      36.64 ms /    55 runs   (    0.67 ms per token,  1501.13 tokens per second)
```
Comment (quernd) [inline org/how-fast-cfg.org:717]: We might want to explicitly contrast these 0.67 both with the 34.63 for structured `llama.cpp` and 0.66 for unstructured, in other words we get *good* structured generation essentially for free, as opposed to *bad* structured generation at a huge run-time cost.

**Comment by quernd** [comment]
Nice writeup. The SQL example is a bit strange as the `SELECT` query returns other columns than `id` for both structured and unstructured generation, but I guess that's beside the point.

>     * [ ]  Determine how much of the interface we want to show (e.g. include/exclude all the index stuff, don't show `ParserLogitsProcessor`, etc.)

Maybe we just mention (for the readers who are entirely unfamiliar with our approach) that we compute the "index" once off-line, but without mentioning construction times for now. The `ParserLogitsProcessor` is in an unexported block. I wouldn't necessarily share it, but maybe emphasize the "secret sauce" of not having to resort to rejection sampling because we have the precise support available. You already mention this by saying "after the entire support (i.e. non-zero “score” tokens) has been determined via our approach" but that could mean many things. One could just enumerate all the tokens (like we do in our tests). We can say something along the lines of "it's just a dict lookup" without giving away the full story about the index structure.

**Comment by rlouf** [comment]
Nice writeup! To add to one of @quernd I think we should add one or two visualizations that help the casual reader see immediately the difference with `llama.cpp`. For instance a GIF that shows generation side by side, and a bar chart that shows the difference in terms of generation time; we could also reproduce the figure we had in the ArXiv paper. I or @willkurt can take care of this.

File: org/how-fast-cfg.org [hunk 3]
```diff
@@ -0,0 +1,778 @@
...
+lp.parse("""
+The following is a SQL statement that returns values from the "id" column of the "users" table:
+```sql
+select id, first_name, last_name, age from users
+```
+""")
+#+END_SRC
+
+#+NAME: outlines-workarounds
+#+BEGIN_SRC python :exports none
```
[the remark this answers is not in the document]
Comment (brandonwillard) [reply org/how-fast-cfg.org:150]: Exactly

**Comment by brandonwillard** [comment]
The code in the `streaming-output` block can be used to produce a side-by-side animated GIF for this post.  I don't know what the right tools are for combining the animations, but I can try to figure something out.  If anyone else is familiar with this sort of thing, don't hesitate to take it from here.

**Comment by brandonwillard** [comment]
All right, a side-by-side animated GIF has been added.

**Comment by brandonwillard** [comment]
> I like the example where you show that structured generation prevents us from generating code that doesn't parse!

Yeah, that was necessary given how likely the underlying model is to produce valid C on its own anyway.

File: org/how-fast-cfg.org (hunk 1 above)
Comment (brandonwillard) [reply org/how-fast-cfg.org:382]: Yes, that's true, but our structured generation will occasionally jump to a millisecond latency for a given sample, so I'm being a little conservative.

**Review by rlouf — APPROVED** [review]

**Pull request merged**